### PR TITLE
auto-detect no_std where possible

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk 
Cc https://rust-lang.zulipchat.com/#narrow/stream/269128-miri/topic/restricted_std.20sysroot.3F

### CODE REVIEW

File: ci.sh [hunk 1]
```diff
@@ -112,7 +112,7 @@ case $HOST_TARGET in
     MIRI_TEST_TARGET=aarch64-linux-android run_tests_minimal hello integer vec panic/panic
     MIRI_TEST_TARGET=wasm32-wasi run_tests_minimal no_std integer strings
     MIRI_TEST_TARGET=wasm32-unknown-unknown run_tests_minimal no_std integer strings
-    MIRI_TEST_TARGET=thumbv7em-none-eabihf MIRI_NO_STD=1 run_tests_minimal no_std # no_std embedded architecture
+    MIRI_TEST_TARGET=thumbv7em-none-eabihf run_tests_minimal no_std # no_std embedded architecture
```
Comment (oli-obk) [inline ci.sh:115]: Weird, why aren't we detecting this with the `contains("-none")` check?

File: cargo-miri/src/setup.rs [hunk 2]
```diff
@@ -67,7 +67,13 @@ pub fn setup(subcommand: &MiriCommand, target: &str, rustc_version: &VersionMeta
...
     };
     // Sysroot configuration and build details.
-    let sysroot_config = if std::env::var_os("MIRI_NO_STD").is_some() {
+    // No-std heuristic taken from rust/src/bootstrap/config.rs
+    // (https://github.com/rust-lang/rust/blob/25b5af1b3a0b9e2c0c57b223b2d0e3e203869b2c/src/bootstrap/config.rs#L549-L555).
+    let no_std = target.contains("-none")
+        || target.contains("nvptx")
+        || target.contains("switch")
+        || target.contains("-uefi");
+    let sysroot_config = if no_std || std::env::var_os("MIRI_NO_STD").is_some() {
```
Comment (oli-obk) [inline cargo-miri/src/setup.rs:76]: Do we want to be able to disable the heuristic with `MIRI_NO_STD=0`?

**Comment by RalfJung** [comment]
Ah, the test suite also checks the env var

https://github.com/rust-lang/miri/blob/defbb668ffe7c28af448cf8054ba100450e8b5b5/tests/compiletest.rs#L97

I don't think I want to repeat the target name check there, that seems a bit too ugly...

**Comment by RalfJung** [comment]
I wonder if we could have ui_test only build the dependencies if at least one test matches the filter? That would elegantly avoid the need for the `MIRI_NO_STD` check.

**Comment by oli-obk** [comment]
Yea, I'm working on this, but it's a bigger refactoring to do it nicely. If you want it fast I can build a hacky version of it.

**Comment by RalfJung** [comment]
Nah, this is not urgent.

**Comment by RalfJung** [comment]
Marking as blocked on https://github.com/rust-lang/miri/pull/3008

**Comment by oli-obk** [comment]
> I wonder if we could have ui_test only build the dependencies if at least one test matches the filter? That would elegantly avoid the need for the `MIRI_NO_STD` check.

this is happening now

**Comment by RalfJung** [comment]
Looks like that is working. :)
@rustbot ready

**Pull request merged**